cdap-master/src/test/java/co/cask/cdap/data/tools/DeletedDatasetMetadataRemoverTest.java

```diff
@@ -0,0 +1,119 @@
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Collections;
+
+/**
+ * Tests for {@link DeletedDatasetMetadataRemover}.
+ */
+public class DeletedDatasetMetadataRemoverTest {
+  @ClassRule
+  public static final DatasetFrameworkTestUtil DS_FRAMEWORK_TEST_UTIL = new DatasetFrameworkTestUtil();
+
+  private static MetadataStore metadataStore;
+  private static DeletedDatasetMetadataRemover metadataRemover;
+
+  @BeforeClass
+  public static void setup() {
+    NamespaceStore nsStore = new InMemoryNamespaceStore();
+    nsStore.create(NamespaceMeta.DEFAULT);
+
+    Injector injector = DS_FRAMEWORK_TEST_UTIL.getInjector().createChildInjector(
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          install(new FactoryModuleBuilder()
+                    .implement(DatasetDefinitionRegistry.class, DefaultDatasetDefinitionRegistry.class)
+                    .build(DatasetDefinitionRegistryFactory.class));
+          bind(DatasetFramework.class)
+            .annotatedWith(Names.named(DataSetsModules.BASIC_DATASET_FRAMEWORK))
+            .to(InMemoryDatasetFramework.class);
+          bind(MetadataChangePublisher.class).to(NoOpMetadataChangePublisher.class);
+          bind(MetadataStore.class).to(DefaultMetadataStore.class);
+        }
+      }
+    );
+    metadataStore = injector.getInstance(MetadataStore.class);
+    metadataRemover = new DeletedDatasetMetadataRemover(nsStore, metadataStore, DS_FRAMEWORK_TEST_UTIL.getFramework());
+  }
+
+  @Test
+  public void test() throws IOException, DatasetManagementException {
+    DatasetId ds1 = NamespaceId.DEFAULT.dataset("ds1");
+    DatasetId ds2 = NamespaceId.DEFAULT.dataset("ds2");
+    DatasetId ds3 = NamespaceId.DEFAULT.dataset("ds3");
+    DS_FRAMEWORK_TEST_UTIL.createInstance("table", ds1.toId(), DatasetProperties.EMPTY);
+    DS_FRAMEWORK_TEST_UTIL.createInstance("table", ds2.toId(), DatasetProperties.EMPTY);
+    DS_FRAMEWORK_TEST_UTIL.createInstance("table", ds3.toId(), DatasetProperties.EMPTY);
+    Assert.assertEquals(3, DS_FRAMEWORK_TEST_UTIL.list(NamespaceId.DEFAULT.toId()).size());
+    metadataStore.addTags(MetadataScope.USER, ds1.toId(), "ds1Tag1", "ds1Tag2");
+    metadataStore.setProperties(MetadataScope.USER, ds2.toId(), Collections.singletonMap("ds2Key", "ds2Value"));
+    metadataStore.addTags(MetadataScope.USER, ds3.toId(), "ds3Tag1", "ds3Tag2");
+
+    verifyMetadataRemoval(ds1);
+    verifyMetadataRemoval(ds2);
+    verifyMetadataRemoval(ds3);
+  }
+
+  private void verifyMetadataRemoval(DatasetId dsId) throws IOException, DatasetManagementException {
+    DS_FRAMEWORK_TEST_UTIL.deleteInstance(dsId.toId());
+    assertNonEmptyMetadata(dsId);
+    metadataRemover.remove();
+    assertEmptyMetadata(dsId);
+  }
+
+  private void assertNonEmptyMetadata(DatasetId dsId) {
+    MetadataRecord metadata = metadataStore.getMetadata(MetadataScope.USER, dsId.toId());
```

Comment:
Its not necessary, because we use ``InMemoryDatasetFramework`` in this test, which does not add any ``SYSTEM`` metadata (or in fact does not do any metadata operations at all)